llpc/lower/llpcSpirvLowerGlobal.cpp

```diff
@@ -370,163 +379,221 @@ static bool hasVertexIdx(const Constant &metaVal) {
 
-    auto inOut = cast<GlobalVariable>(loadSrc);
-    auto inOutTy = inOut->getType()->getContainedType(0);
+  unsigned operandIdx = 0;
+  Value *vertexIdx = nullptr;
+  auto inOutTy = inOut->getType()->getContainedType(0);
 
-    MDNode *metaNode = inOut->getMetadata(gSPIRVMD::InOut);
-    assert(metaNode);
-    auto inOutMetaVal = mdconst::dyn_extract<Constant>(metaNode->getOperand(0));
+  MDNode *metaNode = inOut->getMetadata(gSPIRVMD::InOut);
+  assert(metaNode);
+  auto inOutMetaVal = mdconst::dyn_extract<Constant>(metaNode->getOperand(0));
+
+  // If the input/output is arrayed, the outermost index might be used for vertex indexing
+  if (inOutTy->isArrayTy() && hasVertexIdx(*inOutMetaVal)) {
+    inOutTy = inOutTy->getArrayElementType();
+    vertexIdx = indexOperands[0];
+    ++operandIdx;
+    inOutMetaVal = cast<Constant>(inOutMetaVal->getOperand(1));
+  }
 
-    loadValue = UndefValue::get(inOutTy);
+  Value *loadValue = loadInOutMember(inOutTy, addrSpace, indexOperands, operandIdx, 0, inOutMetaVal, nullptr, vertexIdx,
+                                     InterpLocUnknown, nullptr, &loadInst);
+  m_loadInsts.insert(&loadInst);
+  loadInst.replaceAllUsesWith(loadValue);
+}
 
-    if (inOutTy->isArrayTy() && hasVertexIdx(*inOutMetaVal)) {
-      auto elemTy = inOutTy->getArrayElementType();
-      auto elemMeta = cast<Constant>(inOutMetaVal->getOperand(1));
+// =====================================================================================================================
+// Handle "load" instructions.
+void SpirvLowerGlobal::handleLoadInst() {
+  auto handle = [&](const unsigned addrSpace) {
+    if (addrSpace != SPIRAS_Input && addrSpace != SPIRAS_Output)
+      return false;
+    // Skip if "load" instructions are not expected to be handled
+    const bool isTcsInput = (m_shaderStage == ShaderStageTessControl && addrSpace == SPIRAS_Input);
+    const bool isTcsOutput = (m_shaderStage == ShaderStageTessControl && addrSpace == SPIRAS_Output);
+    const bool isTesInput = (m_shaderStage == ShaderStageTessEval && addrSpace == SPIRAS_Input);
+    return isTcsInput || isTcsOutput || isTesInput;
+  };
 
-      const unsigned elemCount = inOutTy->getArrayNumElements();
-      for (unsigned i = 0; i < elemCount; ++i) {
-        Value *vertexIdx = ConstantInt::get(Type::getInt32Ty(*m_context), i);
-        auto elemValue = addCallInstForInOutImport(elemTy, addrSpace, elemMeta, nullptr, 0, nullptr, vertexIdx,
-                                                   InterpLocUnknown, nullptr, &loadInst);
-        loadValue = InsertValueInst::Create(loadValue, elemValue, {i}, "", &loadInst);
+  for (GlobalVariable &global : m_module->globals())
+    for (User *user : global.users()) {
+      if (LoadInst *loadInst = dyn_cast<LoadInst>(user)) {
+        // The user is a load
+        const unsigned addrSpace = global.getType()->getPointerAddressSpace();
+        if (!handle(addrSpace))
+          continue;
```

Comment:
I think this can be moved outside of the users loop if the global addrspace is the same as the gep addrspace (which I hope is the case?).